feat: Editorial review, Relative color syntax (#32004)
mdn/content: files/en-us/web/css/color_value/hwb/index.md

slug: Web/CSS/color_value/hwb
page-type: css-function
browser-compat: css.types.color.hwb
spec-urls:
  - https://drafts.csswg.org/css-color-5/#relative-HWB
  - https://drafts.csswg.org/css-color/#the-hwb-notation
---

{{CSSRef}}

The **`hwb()`** functional notation expresses a color in the {{glossary("RGB", "sRGB")}} {{glossary("color space")}} according to its hue, whiteness, and blackness. An optional alpha component represents the color's transparency.

{{EmbedInteractiveExample("pages/css/function-hwb.html")}}

## Syntax

```css
/* Absolute values */
hwb(194 0% 0%)
hwb(194 0% 0% / .5)

/* Relative values */
hwb(from green h w b / 0.5)
hwb(from #0000FF h calc(w + 30) b)
hwb(from lch(40%, 70, 240deg) h w calc(b - 30))
```

> > **Note:** The computed values of colors expressed using `hwb()` syntax are {{CSSXref("color_value/rgb", "rgb()")}} values in the case of absolute colors, and {{CSSXref("color_value/color", "color(srgb ... )")}} values in the case of relative colors.

Author: I elected to remove this note, and use the same kind of note as you suggested for `hsl()`, in the same place (at the bottom of the "Relative value syntax" section)

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

> > **Note:** Absolute and relative functional notation serializes to sRGB values, and the values of the red, green, blue components may be rounded in serialization.

Author: Removed

#### Absolute value syntax

```text
hwb(H W B[ / A])
```

The parameters are as follows:

- `H`

  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case) representing the color's {{CSSXref("&lt;hue&gt;")}} angle.

- `W`, `B`

>   - : Each one is a {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). These values represent the color's whiteness and blackness, respectively. They specify the amount of white and black to mix in, from `0%` (no whiteness or blackness) to `100%` (full whiteness or blackness).

Author: I decided to rewrite this a different way:

>   - : Each one is a {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). These values represent the color's whiteness and blackness respectively, specifying the amount to mix in. `0%` represents no whiteness or blackness, whereas `100%` represents full whiteness or blackness.

>     If `W + B = 100%`, it defines a shade of gray. If `W + B > 100%`, `W` and `B` are effectively normalized as `W / (W + B)` and `B / (W + B)`, respectively.

Author: updated

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

#### Relative value syntax

```text
hwb(from <color> H W B[ / A])
```

The parameters are as follows:

- `from <color>`
  - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. The origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `H`
  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case) representing the output color's {{CSSXref("&lt;hue&gt;")}} angle.
- `W`, `B`

>   - : Each value can be written as a {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). These values represent the whiteness and blackness channel values of the output color, respectively. They specify the amount of white and black to mix in, from `0%` (no whiteness or blackness) to `100%` (full whiteness or blackness). If `W + B = 100%`, it defines a shade of gray. If `W + B > 100%`, `W` and `B` are effectively normalized as `W / (W + B)` and `B / (W + B)`, respectively.

Author: updated

```css
hwb(from hsl(0 100% 50%) h w b)
```

This function's output color is the sRGB `color()` equivalent of `hwb(0, 0%, 0%)`: `color(srgb 1 0 0)`.

The next function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css
hwb(from hsl(0 100% 50%) 240 52% 12%)
```

In the above case, the output color is the sRGB `color()` equivalent of `hwb(240 52% 12%)`: `color(srgb 0.52 0.52 0.88)`.

The following function creates a relative color based on the origin color:

```css
hwb(from hsl(0 100% 50%) h 30% b)
```

This example:

- Converts the origin color (`hsl(0 100% 50%)`) into an `hwb()` equivalent (`hwb(0 0% 0%)`).
- Sets the `H` and `B` channel values for the output color to those of the origin color `hwb()` equivalent's `H` and `B` channel values — those values are `0` and `0%`, respectively.
- Sets the output color's `W` channel value to a new value not based on the origin color: `30%`. The output `hwb()` color is `hwb(0 30% 0%)`.

> The final output color is the equivalent color in the sRGB color space — `color(srgb 1 0.3 0.3)`.

Author: Same comment as for the `hsl()` page; I was not that fond of this rewrite. I've updated the final output line to 

> The final output color is the equivalent of `hwb(0 30% 0%)` in the sRGB color space — `color(srgb 1 0.3 0.3)`.

In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.

Let's look at some examples that specify origin and output alpha channel values. The first one specifies the output alpha channel value as being the same as the origin alpha channel value, whereas the second one specifies a different output alpha channel value, unrelated to the origin alpha channel value.

```css
hwb(from hsl(0 100% 50% / 0.8) h w b / alpha)
/* Computed output color: color(srgb 1 0 0 / 0.8) */

hwb(from hsl(0 100% 50% / 0.8) h w b / 0.5)
/* Computed output color: color(srgb 1 0 0 / 0.5) */
```

> In the following example, the `hsl()` origin color is again converted into an `hwb()` representation — `hwb(0 0% 0%)`. {{cssxref("calc")}} calculations are applied to the `H`, `W`, `B`, and `A` values, and the final output color is `color(srgb 0.25 0.9 0.25 / 0.9)`:

Reviewer: ```suggestion
In the following example, the `hsl()` origin color is again converted into an `hwb()` representation — `hwb(0 0% 0%)`. {{cssxref("calc")}} calculations are applied to the `H`, `W`, `B`, and `A` values, and the final output color is the equivalent of `hwb(120 25% 10% / 0.9` : it's `color(srgb 0.25 0.9 0.25 / 0.9)`:
```

Author: updated

```css
hwb(from hsl(0 100% 50%) calc(h + 120) calc(w + 25) calc(b + 10) / calc(alpha - 0.1))
```

> > **Note:** Because the origin color channel values are resolved to `<number>` values, you have to add numbers to them when using them in calculations, even in cases where a channel would normally accept `<percentage>` values or other types. Adding a `<percentage>` to a `<number>`, for example, doesn't work.

Reviewer: ```suggestion
> **Note:** Because the origin color channel values are resolved to `<number>` values, you have to add numbers to them when using them in calculations, even in cases where a channel would normally accept `<percentage>`, `<angle>`, or other value types. Adding a `<percentage>` to a `<number>`, for example, doesn't work.
```

Author: updated

hwb(194 0% 0% / .5) /* #00c3ff with 50% opacity */

/* Relative values */
hwb(from green h w b / 0.5)
hwb(from #0000FF h calc(w + 30) b)
hwb(from lch(40%, 70, 240deg) h w calc(b - 30))
```

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

> **Note:** Absolute and relative functional notation serializes to sRGB values, and the values of the red, green, blue components may be rounded in serialization.

#### Absolute value syntax

```text
hwb(H W B[ / A])
```

The parameters are as follows:

- `H`

  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case). This represents the color's hue angle. More details on this type can be found on the {{CSSXref("&lt;hue&gt;")}} reference.

- `W`, `B`

  - : Each one is a {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). These values represent the color's whiteness and blackness, respectively. They specify the amount of white and black to mix in, from `0%` (no whiteness or blackness) to `100%` (full whiteness or blackness).

>     If `W + B = 100%`, it defines some shade of gray. If `W + B > 100%`, `W` and `B` are effectively normalized as `W / (W + B)` and `B / (W + B)`, respectively.

Author: Updated

- `A` {{optional_inline}}

  - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. If included, the value is preceded by a slash (`/`).

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

#### Relative value syntax

```text
hwb(from <color> H W B[ / A])
```

The parameters are as follows:

- `from <color>`
  - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. Note that the origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `H`
  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case). This represents the hue angle of the output color.
- `W`, `B`

>   - : Each value can be written as a {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). These values represent the whiteness and blackness channel values of the output color, respectively. They specify the amount of white and black to mix in, from `0%` (no whiteness or blackness) to `100%` (full whiteness or blackness). If `W + B = 100%`, it defines some shade of gray. If `W + B > 100%`, `W` and `B` are effectively normalized as `W / (W + B)` and `B / (W + B)`, respectively.

Author: updated

- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

Author: I'm not making this change; see my explanation as to why not in the equivalent change in the `hsl()` page.

#### Defining relative color output channel components

When using relative color syntax inside an `hwb()` function, the browser converts the origin color into an equivalent HWB color (if it is not already specified as such). The color is defined as three distinct color channel values — `h` (hue), `w` (white), and `b` (black) — plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:

- The `h` channel value is resolved to a `<number>` between `0` and `360`.
- The `w` and `b` channels are each resolved to a `<number>` between `0` and `100`.

> - The `alpha` channel is resolved to a `<number>` between `0` and `1`, inclusive.

Author: updated

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

> It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `hwb()` syntax.

Author: updated

Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). The following function outputs the same color as the origin color — it uses the origin color's `h`, `w`, and `b` channel values as the output channel values:

```css
hwb(from hsl(0 100% 50%) h w b)
```

> **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model or space as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels). For example, in the above case the {{cssxref("color_value/hsl", "hsl()")}} color `hsl(0 100% 50%)` is converted to an sRGB `color()` function equivalent to `hwb(0 0 0)`: `color(srgb 1 0 0)`.

This function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css

> hwb(from hsl(0 100% 50%) 240 52% 12%)

Author: Yup, I think you are. See my other comments on the matter. No change made - the output color is specified in the paragraph just after this code block.

```

In the above case, the output color is the sRGB `color()` equivalent of `hwb(240 52% 12%)`: `color(srgb 0.52 0.52 0.88)`.

The following function creates a relative color based on the origin color:

```css
hwb(from hsl(0 100% 50%) h 30% b)
```

This example:

> - Converts the `hsl()` origin color to an equivalent of the `hwb()` representation in the sRGB color space — `color(srgb 1 0 0)` behind the scenes.

Author: OK, I think I've worked out where the confusion lies about what the final output color is, where the conversions happen to output color function equivalent and then to sRGB final output color, etc. I think what happens is this:

1. Origin color is converted into an equivalent color of the specified color function.
2. The channel updates are done according to the values you set in your color function and you get an output color.
3. The final output color (computed value) is the equivalent color in the sRGB color space.

I've just spent ages rewriting these sections to account for this, and I think the result makes a lot more sense than what was there before. Let me know what you think.

> - Sets the `H` and `B` channel values for the output color to those of the origin `color()` equivalent's `H` and `B` channel values — those values are `1` and `0`, respectively.

Reviewer: ```suggestion
- Sets the `H` and `B` channel values for the output color to those of the origin `color()` equivalent's `H` and `B` channel values — those values are both `0`.
```

Author: I've changed this bit to `0` and `0%`.

- Sets the output color's `W` channel value to a new value not based on the origin color: `30%`.

> The above example's output color is `color(srgb 1 0.3 0.3)`.

Reviewer: This is correct, but I don't understand how we got there.

These three are equivalent:
hsl(0 100% 50%) 
hwb(0 0 0)
color(srgb 1 0 0)

if we have `hwb(from hsl(0 100% 50%) h 30% b)`

we have `hwb(from hwb(0 0 0) h 30% b)`.
Replacing the W only with 30%, we get `hwb( 0 30% 0)`
but the output is `color(srgb 1 0.3 0.3)`.

How did the `b`(blue) get to 0.3? 
I do understand how we still have `b` (blackness) of `0` though.

```suggestion
The above example's output color is `hwb(0 0.3 0)`.
```
see https://codepen.io/estelle/pen/RwORbVE

Author: I've changed this bit now anyway (see previous comment), but this makes sense to me - the channels in HWB are not equivalent to the channels in color sRGB. If you add value just to the W channel in an HWB color, it makes the color lighter; the pure sRGB equivalent requires value to be added to both channels that are not full (G and B, in this case) to produce the same lightening effect.

Try this — specify an HWB color, say hwb(0 0% 0%), and then increase its W channel, say to hwb(0 30% 0%).

If you look at the computed values, you'll see them as rgb(255, 0, 0) and rgb(255, 77, 77)

> In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.

Reviewer: ```suggestion
> **Note:** In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.
```

Author: Again, doesn't feel like a note to me

Let's look at some examples that specify origin and output alpha channel values. The first one specifies the output alpha channel value as being the same as the origin alpha channel value, whereas the second one specifies a different output alpha channel value, unrelated to the origin alpha channel value.

```css

> hwb(from hsl(0 100% 50% / 0.8) h w b / alpha)
> /* Computed output color: color(srgb 1 0 0 / 0.8) */
> 
> hwb(from hsl(0 100% 50% / 0.8) h w b / 0.5)
> /* Computed output color: color(srgb 1 0 0 / 0.5) */

Reviewer: ```suggestion
hwb(from hsl(0 100% 50% / 0.8) h w b / alpha)
/* 
  Computed output color: color(srgb 1 0 0 / 0.8) 
  Output color: hwb(0 0 0 / 0.8)
*/

hwb(from hsl(0 100% 50% / 0.8) h w b / 0.5)
/* Computed output color: color(srgb 1 0 0 / 0.5)
  Output color: hwb(0 0 0 / 0.5) */
```

Author: Yeah, I've dealt with this in a different way now

```

> In the following example, the `hsl()` origin color is again converted into an equivalent of the `hwb()` representation in the sRGB color space — `color(srgb 1 0 0)`. {{cssxref("calc")}} calculations are applied to the `H`, `W`, `B`, and `A` values, resulting in an output color of `color(srgb 0.714286 0.714286 0.714286 / 0.9)`:

Reviewer: ```suggestion
In the following example, the `hsl()` origin color is again converted into an equivalent of the `hwb()` representation in the sRGB color space — `color(srgb 1 0 0)` and then to `hwb(0 0 0)`. {{cssxref("calc")}} calculations are applied to the `H`, `W`, `B`, and `A` values, resulting in an output color of `hwb(120 0.25 0.1 / 0.9)` or `color(srgb 0.25 0.9 0.25 / 0.9)`:
```

Reviewer: https://codepen.io/estelle/pen/yLrJBzV?editors=1100

Reviewer 2: `hwb(120 0.25 0.1 / 0.9)` is `color(srgb 0 1 0 / 0.9)`

Author: I've rewritten this bit in my next commit, in a slightly different way

```css

> hwb(from hsl(0 100% 50%) calc(h + 120) calc(w + 25) calc(b + 10) / calc(alpha - 0.1))

Reviewer: ```suggestion
hwb(from hsl(0 100% 50%) calc(h + 120) calc(w + .25) calc(b + .10) / calc(alpha - 0.1))
```
The original creates a grey. This is a green.

Reviewer: The original was impossible to explain mathematically

Author: I'd rather keep the correct specified values here, and point out Chrome's/Safari's problems in the BCD, than write incorrect syntax which will then break when Chrome and Safari fix it in the future.

Reviewer: agreed. I guess we have to go with grey.
Chances of them  changing the spec? because chrome makes more sense than the spec

## See also

- {{CSSXref("&lt;color&gt;")}}: For a list of all color notations

> - [Using relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors)

Reviewer: ```suggestion
- [Using relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors)
- [CSS colors](/en-US/docs/Web/CSS/CSS_colors/) module
```

Author: updated
